helix-core/src/main/java/org/apache/helix/task/TaskStateModelFactory.java

```diff
@@ -176,4 +167,21 @@ protected static RealmAwareZkClient createZkClient(HelixManager manager) {
         new HelixZkClient.ZkConnectionConfig(zkAddress),
         clientConfig.createHelixZkClientConfig().setZkSerializer(new ZNRecordSerializer()));
   }
+
+  private static ScheduledExecutorService createThreadPoolExecutor(HelixManager manager) {
+    RealmAwareZkClient zkClient = createZkClient(manager);
+    ScheduledExecutorService executorService = Executors.newScheduledThreadPool(TaskUtil
+            .getTargetThreadPoolSize(zkClient, manager.getClusterName(), manager.getInstanceName()),
+        new ThreadFactory() {
+          private AtomicInteger threadId = new AtomicInteger(0);
+
+          @Override
+          public Thread newThread(Runnable r) {
+            return new Thread(r, "TaskStateModelFactory-task_thread-" + threadId.getAndIncrement());
+          }
+        });
+    zkClient.close();
+
+    return executorService;
```

Comment:
I think it might be helpful to break this function down to:

1. Do I/O read from ZK (`createZkClient(manager)`, `getTargetThreadPoolSize()`, cache the read value into a local variable), and close the zkClient.
2. Initialize an instance of `executorService` using the read value

You could refactor Step 1 into the constructor (this should be 2-3 lines at most) - this way, `createThreadPoolExecutor` only does what its name says it does - just creating creating an in-memory thread pool, nothing else. This way, we keep private functions modular and the intent is clearer to readers.

Reply:
@narendly Unfortunately, to keep things DRY, the constructor has to use `this()`, and `Call to 'this()' must be first statement in constructor body`. 

The current approach is fine in my opinion: `createThreadPoolExecutor()` does what it needs to do to create the thread pool and cleans up after itself.

Reply:
Can you refactor ZK operations out first? What I am saying is if `Executors` fails for any reason would zkclient be closed properly? I don't see any issues with closing zkclient before we move straight on to thread pool initialization.

Reply:
That's true. I wouldn't be able to refactor it into the constructor, but I can place that I/O first in this helper. Thanks for the tips.